src/sinks/file.rs

```diff
@@ -0,0 +1,316 @@
+        }
+    }
+}
+
+#[typetag::serde(name = "file")]
+impl crate::topology::config::SinkConfig for FileSinkConfig {
+    fn build(&self, acker: Acker) -> Result<(super::RouterSink, super::Healthcheck), String> {
+        let sink = PartitionedFileSink::new(
+            Template::from(&self.path_template[..]),
+            self.close_timeout_secs,
+            self.encoding.clone(),
+        )
+        .stream_ack(acker);
+
+        Ok((Box::new(sink), Box::new(future::ok(()))))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+pub struct PartitionedFileSink {
+    path_template: Template,
+    encoding: Option<BasicEncoding>,
+    close_timeout_secs: Option<u64>,
+
+    partitions: HashMap<PathBuf, EmbeddedFileSink>,
+    last_accessed: HashMap<PathBuf, Instant>,
+    closing: Vec<EmbeddedFileSink>,
+}
+
+impl PartitionedFileSink {
+    pub fn new(
+        path_template: Template,
+        close_timeout_secs: Option<u64>,
+        encoding: Option<BasicEncoding>,
+    ) -> Self {
+        PartitionedFileSink {
+            path_template,
+            encoding,
+            close_timeout_secs,
+            partitions: HashMap::new(),
+            last_accessed: HashMap::new(),
+            closing: Vec::new(),
+        }
+    }
+
+    fn collect_old_files(&mut self) {
+        let mut recently_outdated = vec![];
+        for timeout in self.close_timeout_secs {
+            self.last_accessed.retain(|path, time| {
+                if time.elapsed().as_secs() > timeout {
+                    debug!(message = "removing sink", file = ?path);
+                    recently_outdated.push(path.clone());
+                    false
+                } else {
+                    true
+                }
+            });
+        }
+
+        let mut recently_outdated = recently_outdated
+            .into_iter()
+            .map(|ref path| self.partitions.remove(path).unwrap())
+            .collect();
+
+        //it is easier to empty `closing` buffer and put back not closed sinks, see `poll_close`
+        let mut closing: Vec<EmbeddedFileSink> = self.closing.drain(..).collect();
+        closing.append(&mut recently_outdated);
+
+        for file in closing.into_iter() {
+            self.poll_close(file);
+        }
+    }
+
+    fn poll_close(&mut self, mut file: EmbeddedFileSink) {
+        match file.close() {
+            Err(err) => error!("Error while closing FileSink: {:?}", err),
+            Ok(Async::Ready(())) => debug!("a FileSink closed"),
+            Ok(Async::NotReady) => self.closing.push(file),
+        }
+    }
+}
+
+impl Sink for PartitionedFileSink {
+    type SinkItem = Event;
+    type SinkError = ();
+
+    fn start_send(&mut self, event: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        match self.path_template.render(&event) {
+            Ok(bytes) => {
+                let path = PathBuf::from(String::from_utf8_lossy(&bytes).as_ref());
```

Comment:
I wonder if there is a chance here we are actually double allocating which could be pretty expensive.

Reply:
Yeah, the chance is not big in my opininon, but better to replace `from_utf8_lossy` with `from_utf8` anyway due to explicit error handling (instead of trying to use invalid path).